src/tools/memoize.js

```diff
@@ -0,0 +1,165 @@
+// one will have this special `valueKey` in it.
+var valueKey = {};
+
+// Looks up the memoized value in the `WeakMap` tree
+// based on `args`, returning `undefined` if it's not found.
+var getMemoizedValue = function getMemoizedValue(args) {
+  var valueMap = args.reduce(function (map, arg) {
+    // `map` should *always* be a `WeakMap` or `undefined` at this point.
+    // It's `undefined` if the value hasn't been memoized yet.
+    if (!(map && map.get)) {
+      return undefined;
+    }
+    // Need to ensure `arg` is an object, however since this is
+    // a lookup function we won't try to create an object for it
+    // which would be useless.
+    var argObject = normalizeArg(arg);
+    // This should return a `WeakMap` if there's a cached value
+    // for `argObject` or `undefined` if there isn't one.
+    return map.get(argObject);
+  }, weakCache);
+  // If there is a cached value then `valueMap` will be a `WeakMap`.
+  // If the value hasn't been cached then `valueMap` will be
+  // `undefined`. We'll explicitly return `undefined` in the ternary.
+  return valueMap ? valueMap.get(valueKey) : undefined;
+};
+
+// Creates a `WeakMap` tree that points from each item
+// in the `args` array to `valueToMemoize`.
+// Returns `valueToMemoize`.
+var memoizeValue = function memoizeValue(args, valueToMemoize) {
+  return args.reduce(function (map, arg, idx) {
+    // Get or create an object for `arg` depending
+    // on if it's a primitive value because `WeakMap`
+    // necessitates objects for keys.
+    var argObject = normalizeArg(arg) || setObjectForPrimitive(arg);
+    // If there's no key in `map` for `argObject` yet,
+    // we need to add it.
+    if (!map.has(argObject)) {
+      var isLast = idx === args.length - 1;
+      var newMap = new WeakMap();
+      // If this is the last argument then we need to
+      // associate the value in the `newMap`.
+      if (isLast) {
+        newMap.set(valueKey, valueToMemoize);
+      }
+      map.set(argObject, newMap);
+    }
+    // Should always return a `WeakMap`.
+    return map.get(argObject);
+    // Note that the `|| valueToMemoize` is a workaround for IE11.
+    // Early versions of it won't `set` frozen objects as keys in
+    // `WeakMap`s which results in the above `set` potentially failing
+    // and therefore not returning the value.
+  }, weakCache).get(valueKey) || valueToMemoize;
+};
+
+// Returns a function whose arguments will create a `WeakMap`
+// tree that should self-garbage collect. For example if there
+// are arguments `a`, `b`, and `c`, `a` will point to a `WeakMap`
+// with a key of `b`, which will point to a `WeakMap` with a key
+// of `c`, which will point to the result of `fn`.
+var memoize = function memoize(fn) {
+  return function () {
+    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
+      args[_key] = arguments[_key];
+    }
+
+    return getMemoizedValue(args) || memoizeValue(args, fn.apply(undefined, args));
```

Comment:
@stevenhauser has some fixes to provide here for caching `undefined` - should be able to open source this soon instead of vendoring it.